src/endpoints/authentication/cose_auth.cpp

```diff
@@ -0,0 +1,247 @@
+      }
+
+      if (header_items[ALG_INDEX].uDataType == QCBOR_TYPE_NONE)
+      {
+        throw COSEDecodeError("Missing algorithm in protected header");
+      }
+      if (header_items[KID_INDEX].uDataType != QCBOR_TYPE_NONE)
+      {
+        parsed.kid = qcbor_buf_to_string(header_items[KID_INDEX].val.string);
+      }
+      if (header_items[GOV_MSG_TYPE].uDataType != QCBOR_TYPE_NONE)
+      {
+        parsed.gov_msg_type =
+          qcbor_buf_to_string(header_items[GOV_MSG_TYPE].val.string);
+      }
+      if (header_items[GOV_MSG_PROPOSAL_ID].uDataType != QCBOR_TYPE_NONE)
+      {
+        parsed.gov_msg_proposal_id =
+          qcbor_buf_to_string(header_items[GOV_MSG_PROPOSAL_ID].val.string);
+      }
+      parsed.alg = header_items[ALG_INDEX].val.int64;
+
+      QCBORDecode_ExitMap(&ctx);
+      QCBORDecode_ExitBstrWrapped(&ctx);
+
+      return parsed;
+    }
+
+    bool is_ecdsa_alg(int64_t cose_alg)
+    {
+      return cose_alg == T_COSE_ALGORITHM_ES256 ||
+        cose_alg == T_COSE_ALGORITHM_ES384 ||
+        cose_alg == T_COSE_ALGORITHM_ES512;
+    }
+
+    struct COSESignatureValidationError : public std::runtime_error
+    {
+      COSESignatureValidationError(const std::string& msg) :
+        std::runtime_error(msg)
+      {}
+    };
+  }
+
+  MemberCOSESign1AuthnPolicy::MemberCOSESign1AuthnPolicy() = default;
+  MemberCOSESign1AuthnPolicy::~MemberCOSESign1AuthnPolicy() = default;
+
+  std::unique_ptr<AuthnIdentity> MemberCOSESign1AuthnPolicy::authenticate(
+    kv::ReadOnlyTx& tx,
+    const std::shared_ptr<ccf::RpcContext>& ctx,
+    std::string& error_reason)
+  {
+    const auto& headers = ctx->get_request_headers();
+    const auto content_type_it = headers.find(http::headers::CONTENT_TYPE);
+    if (content_type_it == headers.end())
+    {
+      error_reason =
+        fmt::format("Missing {} header", http::headers::CONTENT_TYPE);
+      return nullptr;
+    }
+    if (content_type_it->second != "application/cose")
+    {
+      error_reason = "Content type is not set to application/cose";
+      return nullptr;
+    }
+
+    auto phdr = cose::decode_protected_header(ctx->get_request_body());
+
+    if (!phdr.kid.has_value())
+    {
+      error_reason = "No kid specified in protected headers";
+      return nullptr;
+    }
+
+    if (!cose::is_ecdsa_alg(phdr.alg))
+    {
+      error_reason = fmt::format("Unsupported algorithm: {}", phdr.alg);
+      return nullptr;
+    }
+
+    MemberCerts members_certs_table(Tables::MEMBER_CERTS);
+    auto member_certs = tx.ro(members_certs_table);
+    auto member_cert = member_certs->get(phdr.kid.value());
+    if (member_cert.has_value())
+    {
+      auto verifier = crypto::make_cose_verifier(member_cert->raw());
+
+      q_useful_buf_c signed_cose;
+      q_useful_buf_c authned_content;
+      signed_cose.ptr = ctx->get_request_body().data();
+      signed_cose.len = ctx->get_request_body().size();
+      if (!verifier->verify(signed_cose, authned_content))
```

Comment:
Following up on the comment from earlier: If this is the only place we're calling this `verify`, I think it should take a `std::span<const uint8_t>` externally, and only convert that to `q_useful_buf_c` internally, removing visibility of the qcbor types.

Reply:
Yes, that's a good point.